posthog/api/github.py

```diff
@@ -0,0 +1,161 @@
+import base64
+import requests
+
+from django.apps import apps
+from django.core.cache import cache
+from django.db.models import Q
+from hashlib import sha256
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+from rest_framework.parsers import JSONParser
+from rest_framework.permissions import AllowAny
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.exceptions import InvalidSignature
+
+from posthog.models.personal_api_key import find_personal_api_key
+from posthog.api.personal_api_key import PersonalAPIKeySerializer
+from posthog.tasks.email import send_personal_api_key_exposed
+
+GITHUB_KEYS_URI = "https://api.github.com/meta/public_keys/secret_scanning"
+TWENTY_FOUR_HOURS = 60 * 60 * 24
+
+
+class SignatureVerificationError(Exception):
+    pass
+
+
+def verify_github_signature(payload: str, kid: str, sig: str) -> None:
+    cache_key = f"github:public_key:{kid}"
+
+    pem = cache.get(cache_key)
+
+    if pem is None:
+        try:
+            resp = requests.get(GITHUB_KEYS_URI, timeout=10)
+            resp.raise_for_status()
+            data = resp.json()
+        except Exception:
+            raise SignatureVerificationError("Failed to fetch GitHub public keys")
+
+        public_keys = data.get("public_keys")
+        if not isinstance(public_keys, list) or not public_keys:
+            raise SignatureVerificationError("No public keys found")
+
+        entry = next((k for k in public_keys if k.get("key_identifier") == kid), None)
+        if entry is None:
+            raise SignatureVerificationError("No public key found matching key identifier")
+
+        pem = entry.get("key")
+        if not isinstance(pem, str) or not pem.strip():
+            raise SignatureVerificationError("Malformed public key entry")
+
+        cache.set(cache_key, pem, TWENTY_FOUR_HOURS)
```

Comment:
Individual keys are cached for 24 hours to prevent spamming the GitHub API. Keys should be immutable so we can make this as long as we'd like.